### ISSUE TITLE
is scheduling Fibers by myself possible ?

### ISSUE THREAD

**Issue by ltearno** [puniverse/quasar#99]
Hello, i need to schedule a fiber by myself, in the way that this is done in the original Matthiass Mann library.

So the code would look like :

	MyScheduler scheduler = new MyScheduler(...);
	Fiber<Void> fiber = new Fiber(scheduler);
	fiber.start();

	while( scheduler.hasSomethingToDo() )
	{
		// executes the scheduled fiber on the current thread
		// if the fiber's code calls 'park' then the scheduler.step() returns
		scheduler.step();
		
		// prepare next things to do...
		// this will eventually unpark some of the previously parked fibers
	}

Is that possible do you think by implementing a new FiberScheduler ?

Note : the thread on which i will execute the fibers 

Thanks for your anwser
And thanks for this great library, I love Quasar !!!

Arnaud Tournier

**Comment by pron** [comment]
It is possible (though I haven't tried), but you really shouldn't do it. It's not our intent to expose continuations directly (we may in the future, though).

However, if you still want to try, try this:

``` java
FiberScheduler myScheduler = new FiberExecutorScheduler("cont", r -> r.run());
```

This would create a scheduler that would schedule tasks to run in-place on the calling thread. Then:

``` java
Fiber<Void> fiber = new Fiber(myScheduler, () -> { ... });
fiber.start();
while(!fiber.isDone()) {
    fiber.unpark();
}
```

**Comment by ltearno** [comment]
Thanks for your answer.

But as far as i understand, in your example, the fiber is not executed in
the caller's thread, but instead it is running in the FiberTimedScheduler's
working thread. And the call to fiber.unpark() will spin until the called
fiber is really unparked - so that's not the behavior i wanted to run.

Am i right ?

That's a good news that you will (maybe) expose the continuations directly
! I am totally for it (as a freelance i might be available to implement
that if you want ;) ).
I must say that the improvments you added to the Matthias Mann library is
very cool to me (especially the ant task detecting all the methods needed
to be instrumented, and the @Suspendable annotation is really cool too). At
the current time, combining Quasar and the original library poses a bit of
problem but i think that's normal (methods instrumented by Matthias Mann
dont get instrumented by Quasar it seems to me).

Thanks a lot

Arnaud

**Comment by pron** [comment]
The fiber should execute inline in the caller's thread. `FiberTimedScheduler` shouldn't be involved at all. But why don't you try and let me know? :)

**Comment by ltearno** [comment]
OK, i will ! Thanks a lot

**Comment by ltearno** [comment]
Ron, i just tried and here are my result. Summary is that your solution
works when called from a normal non-fiber java thread. But when this code
is run inside a fiber, then things go wrong...

Here are the details.

Your source code, with just some sysout to check what happens :

FiberScheduler myScheduler = new FiberExecutorScheduler("cont", r ->
r.run());

Fiber<Void> fiber = new Fiber<Void>(myScheduler, () -> {
System.out.println("Starting");

System.out.println("fiber thread before parking id " +
Thread.currentThread().getId());

Fiber.park();

System.out.println("fiber after unparking thread id " +
Thread.currentThread().getId());

System.out.println("Finished");
});

System.out.println("main thread id " + Thread.currentThread().getId());

System.out.println("starting");
fiber.start();
System.out.println("started");
while (!fiber.isDone())
{
System.out.println("unparking");
fiber.unpark();
System.out.println("unparked");
}






When executed directly in a main method, it will give this result (note
that i am using instrumentation at compile time, that's why the agent is
not here) :




QUASAR WARNING: Quasar Java Agent isn't running. If you're using another
instrumentation method you can ignore this message; otherwise, please refer
to the Getting Started section in the Quasar documentation.
main thread id 1
starting
Starting
fiber thread before parking id 1
started
unparking
fiber after unparking thread id 1
Finished
unparked
Terminated






So here, things are very very cool !



Now, the same code executed inside another fiber :

try
{
new Fiber<Void>(() -> {
FiberScheduler myScheduler = new FiberExecutorScheduler("cont", r ->
r.run());

Fiber<Void> fiber = new Fiber<Void>(myScheduler, () -> {
System.out.println("Starting");

System.out.println("fiber thread before parking id " +
Thread.currentThread().getId());

Fiber.park();

System.out.println("fiber after unparking thread id " +
Thread.currentThread().getId());

System.out.println("Finished");
});

System.out.println("main thread id " + Thread.currentThread().getId());

System.out.println("starting");
fiber.start();
System.out.println("started");
while (!fiber.isDone())
{
System.out.println("unparking");
fiber.unpark();
System.out.println("unparked");
}

}).start().join();
}
catch (ExecutionException e)
{
}
catch (InterruptedException e)
{
}

System.out.println("Terminated");

And the results when executed :

QUASAR WARNING: Quasar Java Agent isn't running. If you're using another
instrumentation method you can ignore this message; otherwise, please refer
to the Getting Started section in the Quasar documentation.
main thread id 13
starting
Starting
fiber thread before parking id 13
started
unparking
Starting
fiber thread before parking id 13
Exception in Fiber "fiber-cont-10000002" java.lang.AssertionError: Illegal
task state: -2
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(ParkableForkJoinTask.java:200)
at
co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park(FiberForkJoinScheduler.java:279)
at co.paralleluniverse.fibers.Fiber.park1(Fiber.java:669)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:582)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:602)
at fr.lteconsulting.RonApp.lambda$2(RonApp.java:23)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run(Unknown Source)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:44)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:32)
at co.paralleluniverse.fibers.Fiber.run(Fiber.java:1019)
WARNING: Uninstrumented methods (marked '**') or call-sites (marked '!!')
detected on the call stack:
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(java.lang.Object,boolean)
(ParkableForkJoinTask.java:200 bci: 73) **
at co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park
(FiberForkJoinScheduler.java:279 bci: 3) **
at co.paralleluniverse.fibers.Fiber.park1 (Fiber.java:669 bci: 114)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:582 bci: 8)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:602 bci: 4)
at fr.lteconsulting.RonApp.lambda$2 (RonApp.java:23 bci: 90)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run (Unknown Source bci: 0)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:44 bci: 4)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:32 bci: 1)
at co.paralleluniverse.fibers.Fiber.run (Fiber.java:1019 bci: 11)
at co.paralleluniverse.fibers.Fiber.run1 (Fiber.java:1014 bci: 1)
Exception in Fiber "fiber-cont-10000002" java.lang.AssertionError: Illegal
task state: -2
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(ParkableForkJoinTask.java:200)
at
co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park(FiberForkJoinScheduler.java:279)
at co.paralleluniverse.fibers.Fiber.park1(Fiber.java:669)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:582)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:602)
at fr.lteconsulting.RonApp.lambda$2(RonApp.java:23)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run(Unknown Source)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:44)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:32)
at co.paralleluniverse.fibers.Fiber.run(Fiber.java:1019)
WARNING: Uninstrumented methods (marked '**') or call-sites (marked '!!')
detected on the call stack:
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(java.lang.Object,boolean)
(ParkableForkJoinTask.java:200 bci: 73) **
at co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park
(FiberForkJoinScheduler.java:279 bci: 3) **
at co.paralleluniverse.fibers.Fiber.park1 (Fiber.java:669 bci: 114)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:582 bci: 8)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:602 bci: 4)
at fr.lteconsulting.RonApp.lambda$2 (RonApp.java:23 bci: 90)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run (Unknown Source bci: 0)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:44 bci: 4)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:32 bci: 1)
at co.paralleluniverse.fibers.Fiber.run (Fiber.java:1019 bci: 11)
at co.paralleluniverse.fibers.Fiber.run1 (Fiber.java:1014 bci: 1)
Exception in Fiber "fiber-10000001" java.lang.AssertionError: Illegal task
state: -2
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(ParkableForkJoinTask.java:200)
at
co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park(FiberForkJoinScheduler.java:279)
at co.paralleluniverse.fibers.Fiber.park1(Fiber.java:669)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:582)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:602)
at fr.lteconsulting.RonApp.lambda$2(RonApp.java:23)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run(Unknown Source)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:44)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:32)
at co.paralleluniverse.fibers.Fiber.run(Fiber.java:1019)
WARNING: Uninstrumented methods (marked '**') or call-sites (marked '!!')
detected on the call stack:
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(java.lang.Object,boolean)
(ParkableForkJoinTask.java:200 bci: 73) **
at co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park
(FiberForkJoinScheduler.java:279 bci: 3) **
at co.paralleluniverse.fibers.Fiber.park1 (Fiber.java:669 bci: 114)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:582 bci: 8)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:602 bci: 4)
at fr.lteconsulting.RonApp.lambda$2 (RonApp.java:23 bci: 90)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run (Unknown Source bci: 0)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:44 bci: 4)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:32 bci: 1)
at co.paralleluniverse.fibers.Fiber.run (Fiber.java:1019 bci: 11)
at co.paralleluniverse.fibers.Fiber.run1 (Fiber.java:1014 bci: 1)
Exception in Fiber "fiber-10000001" java.lang.AssertionError: Illegal task
state: -2
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(ParkableForkJoinTask.java:200)
at
co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park(FiberForkJoinScheduler.java:279)
at co.paralleluniverse.fibers.Fiber.park1(Fiber.java:669)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:582)
at co.paralleluniverse.fibers.Fiber.park(Fiber.java:602)
at fr.lteconsulting.RonApp.lambda$2(RonApp.java:23)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run(Unknown Source)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:44)
at
co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run(SuspendableUtils.java:32)
at co.paralleluniverse.fibers.Fiber.run(Fiber.java:1019)
Terminated
WARNING: Uninstrumented methods (marked '**') or call-sites (marked '!!')
detected on the call stack:
at
co.paralleluniverse.concurrent.forkjoin.ParkableForkJoinTask.park(java.lang.Object,boolean)
(ParkableForkJoinTask.java:200 bci: 73) **
at co.paralleluniverse.fibers.FiberForkJoinScheduler$FiberForkJoinTask.park
(FiberForkJoinScheduler.java:279 bci: 3) **
at co.paralleluniverse.fibers.Fiber.park1 (Fiber.java:669 bci: 114)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:582 bci: 8)
at co.paralleluniverse.fibers.Fiber.park (Fiber.java:602 bci: 4)
at fr.lteconsulting.RonApp.lambda$2 (RonApp.java:23 bci: 90)
at fr.lteconsulting.RonApp$$Lambda$3/1992013462.run (Unknown Source bci: 0)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:44 bci: 4)
at co.paralleluniverse.strands.SuspendableUtils$VoidSuspendableCallable.run
(SuspendableUtils.java:32 bci: 1)
at co.paralleluniverse.fibers.Fiber.run (Fiber.java:1019 bci: 11)
at co.paralleluniverse.fibers.Fiber.run1 (Fiber.java:1014 bci: 1)

So here, i need to invest more time to investigate where the problem
fundamentally comes from. Maybe you have an idea. On the other side, i
completely understand that Quasar was not architectured to handle those
kind of funny edge cases....

Thanks again !

Arnaud

**Comment by pron** [comment]
The problem is that there is there is one ThreadLocal per thread holding the identity of the currently running fiber, and as the code stands, it doesn't allow nesting (the TL's value isn't saved and restored). Why do you need this behavior at all? When needing to run a lot of fibers in lockstep, I simply use a Phaser. This way, you can also enjoy parallelism.

**Comment by ltearno** [comment]
Now i don't need to do that anymore, because i discovered that you already
programmed the ProxyServerActor, which was what i was about to implement.
What i wanted was to be able to call a method on an interface in a blocking
fashion while proxying the call to another fiber. Your ProxyServerActor
does the thing perfectly for now !

In one word i wanted to have many fibers, each one running a message
pumping loop which would be able to process multiple messages in parallel
(each time a message processing function starts a blocking operation, the
stack trace would be rollbacked (waiting the result to come) and another
message could be processed. All that with a synchronous way of coding.

Also since i like technical things, i was also interrested how to do it.
And some times ago i developped a library in C/C++ for x86/x64 which does
binary instrumentation to catch procedure calls. So you guessed it : i like
playing with stack frames and was inspired to see how it was done in java !

And on a practical side, i am developping a server application and wanted
to go out of the classic JEE and make something efficient to manage
concurrency and data sharing.

Thanks again. I hope i didn't take too much of your time. Glad to have
talked with you, people able to write such code as Quasar are not many on
earth (sadly) !

I don't know the Phaser, i will look at it.

And from your reply, i am guessing that to make the thing work, i would
just have to store and reset the TLS before the continuation call and
restore it after... I will try that maybe ...

Thanks !

Arnaud

**Comment by pron** [comment]
What you want can be easily achieved with the single-threaded scheduler I mentioned above.
The TL problem is in the `Fiber` class. You can't fix it from the outside (at least not without reflection).

Thank you very much for your kind words, and I'm very glad you like Quasar!

**Comment by ltearno** [comment]
Yes i am conscious that the TL fields are declared and used in Quasar, so i
may make me a fork ;)

something like : Object storeAndClearFiberLS() and restoreFiberLS( Object
state ) should be enough..

Thanks !
